Use SoC temperature sensor

GMagician opened the pull request:
This PR resume #23765. I added more SoCs than original PR and fixed some bugs in it.

There is something that needs more coding. Some SoCs of the same family may use different calibration registers and/or different calibration temperatures.

GMagician committed:
SoC Temperature Sensor

GMagician committed:
Prevent BOARD_PIN definition

GMagician commented:
I tested code with my S6 V2 that sits inside a box with a thermistor that reads temperature.

There is a difference of about 11°C between the sensors (SoC temp is bigger as expected) when turned off. I'm expecting that this difference increase over time

GMagician commented:
There is no possibility to detect chip inside a family with a so precise degree of correctness. Some chips will not work as expected. In "internal_temp.h" there is a list of chips that may not work or work changing some values (these values are commented)

thinkyhead committed:
Delete [ \t]+$

thinkyhead committed:
misc. cleanup

thinkyhead commented:
I had to change the logic for `TEMP_SENSOR_BOARD` in `Configuration.h` because `TEMP_BOARD_PIN` is not defined by `pins.h` until _after_ `Configuration_adv.h` is already processed.

Along with that I had to alter the define of `TEMP_BOARD_PIN` in the pins files to make it overridable.

For the ERYONE Ery32 mini I added a warning that the `TEMP_SENSOR_BOARD` should be set to 13 when using `TEMP_BOARD_PIN PC3`.

GMagician commented:
I added a comment but I just see "pending" review without any possibility to see message content. Not sure anyone else can read it

thinkyhead commented:
> I added a comment but I just see "pending" review without any possibility to see message content. Not sure anyone else can read it

Delete it and try again. Add it as a regular comment, not a "review comment."

GMagician commented:
Yep, but no button to add comment, just review...

GMagician commented on Marlin/src/HAL/STM32/inc/SanityCheck.h:
```diff
@@ -51,6 +51,10 @@
   #error "TFT_COLOR_UI, TFT_LVGL_UI and TFT_CLASSIC_UI are currently only supported on STM32H7, STM32F4 and STM32F1 hardware."
 #endif
 
+#if TEMP_SENSOR_BOARD_IS_INTERNAL && defined(ATEMP) && TEMP_BOARD_PIN != ATEMP
```
Maybe
```
#if TEMP_SENSOR_BOARD_IS_INTERNAL
  #if !defined(ATEMP) 
    #error "ATEMP is mandatory"
  #else if TEMP_BOARD_PIN != ATEMP
```
because ATEMP is required by code and by message itself

thinkyhead replied:
The only reason I added that is because in some unknown future `ATEMP` could be defined as pin 0.

GMagician replied:
ok but 0 is still "defined" to test for invalid 0 it should look like:
```#if TEMP_SENSOR_BOARD_IS_INTERNAL && ATEMP && TEMP_BOARD_PIN != ATEMP```

thinkyhead committed:
heading cleanup

thinkyhead commented:
> Yep, but no button to add comment, just review...

Sometimes it's there and sometimes it's not.

thisiskeithb commented:
Instead of using the MCU temperature as a board sensor, can it be its own thing? like `TEMP_SENSOR_MCU`?

GMagician commented:
Yes it can, but would you display it somewhere or just used as thermal protection?

thisiskeithb commented:
> Yes it can, but would you display it somewhere or just used as thermal protection?

Since the MCU temperature is not as accurate as a physical thermistor attached to the PCB, just display. I'd be leary of using it for thermal protection due to inaccuracies.

> Because as board temp you see it on hosts

I'm sure hosts will pick up the feature as time goes on.

GMagician commented:
> Since the MCU temperature is not as accurate as a physical thermistor attached to the PCB, just display. I'd be leary of using it for thermal protection due to inaccuracies.

It shouldn't be inaccurate if you consider it as SoC temperature. When you want it to be your "board temperature" then you're always on safe side because SoC is hotter then the board ambient itself (mosfet/stepper drivers excluded of course, but when you use a thermistor, these are not checked as well)
If this sensor has to be turned to "SoC only" then code has to be completely changed, starting from coming a simple enabled/disabled feature.
Please let me know if I have to pursue such path.

thisiskeithb commented:
> If this sensor has to be turned to "SoC only" then code has to be completely changed, starting from coming a simple enabled/disabled feature.

I know it’s more work, but it would be more in line with RRF and Klipper to have a separate MCU temperature.

GMagician committed:
Standalone SoC sensor

GMagician commented:
Now some thing about new commit (not yet tested because printer is busy):
a) HAS_ADC_TEST(SOC) fail because pin is not defined yet
b) sensor protection is done only when temp > setpoint
c) temperature report send S:c/t but target is always 0. Should this follow other temp format?
d) same as c for board temperature
e) TEMP_SENSOR_SOC should be any number, it doesn't use nay table. Must be changed to something different? Id did this way to keep a similar approach as other sensors

GMagician commented:
> Sometimes it's there and sometimes it's not.

It reappeared once closed review!

GMagician committed:
Add missing update temp

GMagician commented:
Ok I tested code. Host report temperature and thermal protection is working. I can't test controller fan because I don't have it.

About SOC_PIN, maybe it may be defined into each board. This way it can be just enabled by what is now TEMP_SENSOR_SOC (or whatever will take its place) and may be added to boards how have correct STM32 supported SoC.

thinkyhead committed:
rename

thinkyhead committed:
Allow config / pins to set TEMP_SOC_PIN

thinkyhead committed:
misc. adjustments, add test

GMagician commented on Marlin/src/pins/pins_postprocess.h:
```diff
@@ -456,6 +456,11 @@
 #endif
 
+// Use ATEMP if TEMP_SOC_PIN is not defined
+#if !defined(TEMP_SOC_PIN) && defined(ATEMP)
```
ATEMP is defined in STM32 environment and maybe not in others. Isn't better to define TEMP_SOC_PIN directly in each MB pins definition files? I know it's a big task but it seems logic since this may be seen as an internal pin and doing that you may define it or not depending on SoC used by the board itself.

thinkyhead replied:
>  Isn't better to define `TEMP_SOC_PIN` directly in each MB pins definition files?

If the platform provides the MCU's internal pin number as `ATEMP` then we don't have to define it in individual pins files. Users are encouraged to set custom pins in their configurations rather than edit pins files, so repeating this item in many pins files offers no special benefit for configuration, or indeed even for board documentation. Presumably all boards with certain MCUs will have an `ATEMP` so it's not a board-related feature, _per se_.

GMagician replied:
yep but not all SoCs have the registers to properly read internal temperature, and there is no way (I found) to properly discriminate SoC inside its family. Defining it into pins may be a solution. It will be defined only for boards that uses the compatible SoCs and users may be free to change it as they want (as it's done for all other pins). SAMD51 has 2 sensors then another decision should be taken (ok, it's not supported there but it is just a way to show different platforms behaviour)

GMagician committed:
Add missing sanity check

GMagician committed:
Add temp_sensor_soc define

Added define but it's really right way to do it or is it better to use a enabled/disabled option?

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25642
